snorkel/learning/pytorch/rnn/lstm.py

```diff
@@ -43,9 +43,16 @@ def forward(self, X, hidden_state):
         seq_lengths, perm_idx = seq_lengths.sort(0, descending=True)
         X = X[perm_idx, :]
         inv_perm_idx = torch.tensor([i for i, _ in sorted(enumerate(perm_idx), key=lambda idx: idx[1])], dtype=torch.long)
-
         encoded_X = self.embedding(X)
-        encoded_X = pack_padded_sequence(encoded_X, seq_lengths, batch_first=True)
+
+        if self.gpu:
+            hidden_state = (hidden_state[0].permute(1,0,2), hidden_state[1].permute(1,0,2))
```

Comment:
@danich1 why the permutation here (and below)?  Add a comment perhaps?

Reply:
Long story short this is a trick to get dataparallel to work with the hidden state and cell state tensors. The dataparallel functions splits on the first dimension, which is implied to be the batch size dimension. This is correct for the input, but for the LSTM module it is a problem. The module expects both tensors to be in the form of ```num layers* num directions, batch_size, dimensions```, which will cause data parallel to split the data on the wrong dimension. (eventually leading to an error) The easiest solution is to permute the hidden state and cell state before the forward call and then permute it back in the foward function. More explanation in this [forum post](https://discuss.pytorch.org/t/multi-layer-rnn-with-dataparallel/4450/5).